### PR TITLE
module_OOP

### PR DESCRIPTION
**Pull request by YaroslavNehutorov**

### CODE REVIEW

File: models.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        if choice == "start":
+            break
+        continue
+    rules()
+    return Player(user_name)
+
+
+def get_enemy_instance(lvl=1) -> Enemy:
+    return Enemy(lvl)
+
```
Comment (shorodilov) [inline models.py]: This should really be moved into the game engine. The `models` module has no responsibility for this stuff.

File: models.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+def get_scores(player_obj: Player) -> None:
+    with open("scores.txt", "a") as score:
+        score.write(f"Player - {player_obj.name}, scored: {player_obj.score} points\n")
```
Comment (shorodilov) [inline models.py]: Move this functions to the `game` module.

File: settings.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    | Available cards: Warrior, Robber, Wizard.      |
+    |        The Wizard defeats the Warrior.         |
+    |        The Warrior defeats the Robber.         |
+    |        The Robber defeats the Wizard.          |
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    |                 Key 1 - Warrior                |
+    |                 Key 2 - Robber                 |
+    |                 Key 3 - Wizard                 |
+    +------------------------------------------------+""")
+
```
Comment (shorodilov) [inline settings.py]: Do not include functions into `constants` unless their help is needed.

File: models.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    def fight(attack, defence) -> int:
+        if attack - defence in exodus_1:
+            return 1
+        elif attack - defence in exodus_2:
+            return -1
+        else:
+            return 0
+
+    def decrease_lives(self) -> None:
+        if self.lives != 1:
```
Comment (shorodilov) [inline models.py]: Let assume `lives` value is set to `1`, this may become a problem.

File: models.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            self.lives -= 1
+            print(f"You have {self.lives} lives left\n")
+        else:
+            print(f"You scored: {self.score}")
+            raise GameOver
+
+    def attack(self, enemy_obj: Enemy) -> None:
+        while True:
```
Comment (shorodilov) [inline models.py]: Do we actually need this endless-loop here?

File: models.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+                print("You missed\n")
+                self.decrease_lives()
+            elif outcome == 1:
+                print("You attacked successfully !\n")
+                enemy_obj.decrease_lives()
+                self.score += 1
+            break
+
+    def defence(self, enemy_obj: Enemy) -> None:
+        while True:
```
Comment (shorodilov) [inline models.py]: Do we actually need this endless-loop here.

File: models.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    def __init__(self, level: int):
+        self.level = level
+        self.lives = level
+
+    @staticmethod
+    def select_attack() -> int:
+        return random.randint(1, 3)
+
+    def decrease_lives(self) -> None:
+        if self.lives != 0:
```
Comment (shorodilov) [inline models.py]: That if `lives` value is set to `0`?

File: models.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
+"""CLASS PLAYER AND ENEMY FOR GAME"""
+
+import random
+from settings import LIVES, exodus_1, exodus_2, rules
```
Comment (shorodilov) [inline models.py]: Do we need `rules` import here?

File: models.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+    def attack(self, enemy_obj: Enemy) -> None:
+        while True:
+            choice_attack = int(input("Choose who you want to attack: "))
+            outcome = self.fight(choice_attack, enemy_obj.select_attack())
+            if outcome == 0:
+                print("It's a draw\n")
+            elif outcome == -1:
+                print("You missed\n")
+                self.decrease_lives()
```
Comment (shorodilov) [inline models.py]: Not included into requirements

File: models.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+            break
+
+
+def get_player_instance() -> Player:
+    user_name = input("Enter your name: ")
+    while True:
+        choice = input("Enter start for start the game: ")
+        if choice == "start":
+            break
+        continue
```
Comment (shorodilov) [inline models.py]: I'll get the same behavior without `continue`.

File: models.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    rules()
```
Comment (shorodilov) [inline models.py]: Does this expected?

File: models.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+                self.decrease_lives()
+            elif outcome == -1:
+                print("You attacked successfully !\n")
+                enemy_obj.decrease_lives()
+                self.score += 1
+            break
+
+
+def get_player_instance() -> Player:
+    user_name = input("Enter your name: ")
```
Comment (shorodilov) [inline models.py]: No `user_name` validation. We should really check it's correct.

File: exeptions.py [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+
+class GameOver(Exception):
+    pass
+
+
+class EnemyDown(Exception):
+    pass
+
+
+class Exit(Exception):
```
Comment (shorodilov) [inline exeptions.py]: This one is useless, just remove it since you're not going to use it in any of your modules.

**Review by shorodilov — COMMENTED** [review]
All loops require a refactoring.
The `Player.defence` method has unexpected side-effects.
This works as a miracle.

File: models.py (hunk 4 above)
Comment (sgorodilov) [reply models.py:40]: It's about `continue` in the end of the loop-block

File: models.py [hunk 14]
```diff
@@ -15,13 +15,15 @@ def select_attack() -> int:
         return random.randint(1, 3)
 
     def decrease_lives(self) -> None:
-        if self.lives != 0:
+
+        if self.lives > 0:
```
Comment (shorodilov) [inline models.py]: Just use `self.lives -= 1` and perform positive validation as `if self.lives < 1`

File: models.py [hunk 15]
```diff
@@ -31,65 +33,83 @@ def __init__(self, name):
...
+        if outcome == 0:
+            print("It's a draw\n")
+
+        elif outcome == -1:
+            print("You missed\n")
+
+        elif outcome == 1:
+            print("You attacked successfully !\n")
+            enemy_obj.decrease_lives()
+            self.score += 1
```
Comment (shorodilov) [inline models.py]: Do not include the most expected behavior into `else` block. Move it to `if`.

File: models.py [hunk 16]
```diff
@@ -31,65 +33,83 @@ def __init__(self, name):
...
+            print("It's a draw\n")
+
+        elif outcome == 1:
+            print("You missed\n")
+            self.decrease_lives()
+
+        elif outcome == -1:
+            print("You attacked successfully !\n")
+            enemy_obj.decrease_lives()
+            self.score += 1
```
Comment (shorodilov) [inline models.py]: This method has a score-change side-effect which is not described in the original scope.
As well move the most expected behavior to the `if` block, do not hide it down side.

File: models.py [hunk 17]
```diff
@@ -31,65 +33,83 @@ def __init__(self, name):
...
+    user_name = check_name()
     while True:
         choice = input("Enter start for start the game: ")
         if choice == "start":
-            break
-        continue
-    rules()
-    return Player(user_name)
-
+            rules()
```
Comment (shorodilov) [inline models.py]: Remove all the unused logics. It a trash functions without any useful payload.

File: models.py [hunk 18]
```diff
@@ -31,65 +33,83 @@ def __init__(self, name):
...
 
-
-def get_scores(player_obj: Player) -> None:
-    with open("scores.txt", "a") as score:
-        score.write(f"Player - {player_obj.name}, scored: {player_obj.score} points\n")
+def check_name():
+    while True:
+        user_name = input("Enter your name: ")
+        if len(user_name) > 2 and user_name.isalpha():
+            return user_name.capitalize()
```
Comment (shorodilov) [inline models.py]: Why it should be capitalized?

File: models.py [hunk 19]
```diff
@@ -31,65 +33,83 @@ def __init__(self, name):
...
-    return Enemy(lvl)
 
-
-def get_scores(player_obj: Player) -> None:
-    with open("scores.txt", "a") as score:
-        score.write(f"Player - {player_obj.name}, scored: {player_obj.score} points\n")
+def check_name():
+    while True:
+        user_name = input("Enter your name: ")
+        if len(user_name) > 2 and user_name.isalpha():
```
Comment (shorodilov) [inline models.py]: Why it should be 3-characters long? I'd like it to be 5 or more chars.

File: models.py [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+
+
+def check_name():
+    while True:
+        user_name = input("Enter your name: ")
+        if len(user_name) > 2 and user_name.isalpha():
+            return user_name.capitalize()
+
+
+def rules():
```
Comment (shorodilov) [inline models.py]: Why this is implemented within `models`?

File: models.py [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+def get_player_instance() -> Player:
+    user_name = check_name()
+    while True:
+        choice = input("Enter start for start the game: ")
+        if choice == "start":
+            rules()
+            return Player(user_name)
+
+
+def check_name():
```
Comment (shorodilov) [inline models.py]: Why this is implemented with `models` module?

File: game.py [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+        try:
+            player.attack(enemy)
+            player.defence(enemy)
+
+        except EnemyDown:
+            player.score += 5
+            enemy = get_enemy_instance(enemy.level + 1)
+
+        except GameOver:
+            def get_scores(player_obj: Player) -> None:
```
Comment (shorodilov) [inline game.py]: And... that if just delete the inner function and handle score-write without it?

**Review by shorodilov — CHANGES_REQUESTED** [review]
The program implementations has side-effects which should be fixed.
While loops should be reviewed.

File: game.py [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+    with open("scores.txt", "a") as score:
+        score.write(f"Player - {player_obj.name}, scored: {player_obj.score} points\n")
+
+
+def get_player_instance() -> Player:
+    user_name = input("Enter your name: ")
+    while not user_name:
+        user_name = input("Enter your name")
+    choice = input("Enter start for start the game: ")
+    if choice == "start":
```
Comment (shorodilov) [inline game.py:15]: Move out from the function. This will break the logic and potentially will raise an exception.

File: models.py [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+        else:
+            print("It's a draw\n")
+
+    def defence(self, enemy_obj: Enemy) -> None:
+        choice_attack = int(input("Choose who you want to protect: "))
+        outcome = self.fight(enemy_obj.select_attack(), choice_attack)
+
+        if outcome == -1:
+            print("You attacked successfully !\n")
+            self.score += 1
```
Comment (shorodilov) [inline models.py:70]: Score assignment is unexpected.

File: game.py [hunk 25]
```diff
@@ -0,0 +1,66 @@
+"""MAIN LOGIC FOR GAME"""
+from models import *
+
+
+def get_scores(player_obj: Player) -> None:
```
Comment (shorodilov) [inline game.py:5]: This should be called `write_scores`

**Review by shorodilov — APPROVED** [review]
There are few issues and output text is non informative.
But it will work in a specif cases.

**Pull request merged**